File: python-main-Underwater Acoustic Channels Group-lizhengnan/src/uwa_channels/replay.py
import numpy as np
from scipy.interpolate import CubicSpline
import scipy.signal as sg
from fractions import Fraction


def replay(input, fs, array_index, channel, start=None):
    """
    Simulate the replay of a passband signal through an underwater acoustic channel.

    Handles three tracking modes:
      theta_hat (phase tracking only):
        h_hat already contains drifting taps. Only phase is re-inserted
        via exp(+j*theta_hat). No delay interpolation is performed.
      phi_hat (delay tracking):
        h_hat is static. Both phase and delay drift are re-inserted.
        Delay drift: drift = phi_hat / (2*pi*fc).
      No tracking:
        Pure convolution with h_hat. Only valid for static channels.

    Parameters:
    -----------
    input : ndarray
        Real passband input signal.

    fs : float
        Sampling frequency of the input signal in Hz.

    array_index : list of int
        Indices of the array elements to simulate.

    channel : dict
        Dictionary containing channel characteristics. Expected keys:

        - "h_hat" : dict
            - "real" : ndarray
            - "imag" : ndarray
        - "params" : dict
            - "fs_delay", "fs_time", "fc" : float (wrapped in np.array)
        - "theta_hat" : ndarray, optional
            Phase estimates (phase tracking only, no delay compensation).
        - "phi_hat" : ndarray, optional
            Phase estimates (delay tracking: phase + delay compensation).
        - "f_resamp" : float, optional
            Factor for additional passband resampling.

    start : int, optional
        Starting index for signal propagation. If None, a random point is chosen.

    Returns:
    --------
    ndarray
        Simulated replay output with dimensions (samples, array_elements).

    Revision history:
      - Apr.  1, 2025: Initial release.
      - Feb. 27, 2026: Fixed delay insertion for theta_hat vs phi_hat.
    """

    # Validate inputs
    validate_inputs(input, fs, array_index, channel)

    # Unpacking variables
    h_hat_real = np.array(channel["h_hat"]["real"])[:, array_index, :]
    h_hat_imag = np.array(channel["h_hat"]["imag"])[:, array_index, :]
    fs_delay = channel["params"]["fs_delay"][0, 0]
    fs_time = channel["params"]["fs_time"][0, 0]
    fc = channel["params"]["fc"][0, 0]
    M = len(array_index)
    L = h_hat_real.shape[2]

    # Convert to baseband and resample the signal to fs_delay
    frac = Fraction(fs_delay / fs).limit_denominator()
    baseband = input * np.exp(-2j * np.pi * fc * np.arange(input.shape[0]) / fs)
    baseband = sg.resample_poly(baseband, frac.numerator, frac.denominator)
    T = baseband.shape[0]

    # Assign random start point in time
    T_max = h_hat_real.shape[0] / fs_time * fs_delay * 1.0
    if start is None:
        start = np.random.randint(low=0, high=T_max - T - L)
    print(f"Start = {start}")

    # Convolution
    buffer = np.zeros((L - 1,))
    baseband = np.concatenate((buffer, baseband, buffer))
    output = np.zeros((T + L, M), dtype=complex)
    channel_time = np.arange(h_hat_real.shape[0]) / fs_time
    signal_time = np.arange(start, start + T + L) / fs_delay
    for m in range(M):
        ir_real = CubicSpline(channel_time, np.squeeze(h_hat_real[:, m, ::-1]))(
            signal_time
        )
        ir_imag = CubicSpline(channel_time, np.squeeze(h_hat_imag[:, m, ::-1]))(
            signal_time
        )
        ir = ir_real + 1j * ir_imag

        if "phi_hat" in channel:
            # Delay tracking: phase + delay re-insertion
            phi_hat = np.array(channel["phi_hat"])[:, array_index[m]]
            for t in np.arange(T + L - 1):
                output[t, m] = (ir[t, :] @ baseband[t : t + L]) * np.exp(
                    1j * phi_hat[t + start]
                )
            drift = phi_hat[np.arange(start, start + T + L)] / (2 * np.pi * fc)
            output[:, m] = CubicSpline(signal_time, output[:, m])(signal_time + drift)
        elif "theta_hat" in channel:
            # Phase tracking only: no delay interpolation
            theta_hat = np.array(channel["theta_hat"])[:, array_index[m]]
            for t in np.arange(T + L - 1):
                output[t, m] = (ir[t, :] @ baseband[t : t + L]) * np.exp(
                    1j * theta_hat[t + start]
                )
        else:
            # No tracking: pure convolution
            for t in np.arange(T + L - 1):
                output[t, m] = ir[t, :] @ baseband[t : t + L]

    # Resample to match the original sampling rate and upshift to fc
    output = sg.resample_poly(output, frac.denominator, frac.numerator)
    output = np.real(
        output * np.exp(2j * np.pi * fc * np.arange(len(output))[:, None] / fs)
    )

    # Resample in passband if needed
    if "f_resamp" in channel:
        frac_resample = Fraction(channel["f_resamp"][0, 0]).limit_denominator()
        output = sg.resample_poly(
            output, frac_resample.numerator, frac_resample.denominator
        )

    output /= np.sqrt(np.sum(pwr(output)))

    return output


def pwr(x):
    return np.mean(np.abs(x) ** 2, axis=0)


def validate_inputs(input, fs, array_index, channel):
    assert (
        channel["version"][0, 0] >= 1.0
    ), f"The minimum version of the channel matrix is 1.0, and you have {channel['version']:.1f}."

    T = input.shape[0]
    T = T / fs
    T_max, N, _ = channel["h_hat"]["real"].shape
    T_max = T_max / channel["params"]["fs_time"][0, 0]

    assert (
        T < T_max
    ), f"Duration of the input signal, {T * 1e3:.2f}ms, should be no larger than {T_max * 1e3:.2f}ms."

    assert len(set(array_index)) == len(array_index), "index contains duplicate values."
    assert (
        max(array_index) <= N
    ), f"array_index must be positive integers and cannot exceed {N}."
    assert input.ndim == 1, "The maximum supported number of channels is 1."


# [EOF]

File: matlab-main/src/replay.m
function output = replay(input, fs, array_index, channel, varargin)
% REPLAY Passes a passband signal through an underwater acoustic channel.
%
% OUTPUT = REPLAY(INPUT, FS, ARRAY_INDEX, CHANNEL) returns the received OUTPUT
% signal after the INPUT passband signal is processed through the CHANNEL.
% FS is the sampling frequency of the INPUT signal, ARRAY_INDEX specifies the 
% hydrophone indices, and CHANNEL is a struct containing the estimated 
% channel coefficients and other relevant parameters.
%
% OUTPUT = REPLAY(INPUT, FS, ARRAY_INDEX, CHANNEL, START) specifies the START 
% index of the channel traces for reproducibility.
%
% Inputs:
%    input               - Input passband signal (time-by-array).
%    fs                  - Sampling frequency of the input signal in Hz.
%    array_index         - Indices of the hydrophones.
%    channel             - Struct containing parameters and impulse responses.
%    varargin            - Optional parameter specifying the start time index.
%
% Outputs:
%    output              - Processed output signal.
%
% Example:
%    See example_replay.m.
%
% Other m-files required: None
% Subfunctions: validate_inputs, pwr
% Toolbox required: Signal Processing Toolbox (for resample function).
% MAT-files required: Channel MAT-file.
%
% See also: noisegen.m
%
% Author: Zhengnan Li
% Email : uwa-channels@ofdm.link
%
% License: MIT
%
% Revision history:
%   - Apr. 1, 2025: Initial release.
%

%% Simple checks
validate_inputs(input, fs, array_index, channel);

%% Unpacking variables
fs_delay = channel.params.fs_delay; % Sampling rate in delay domain.
fs_time = channel.params.fs_time; % Sampling rate in time.
fc = channel.params.fc; % Channel center frequency.
M = length(array_index); % Number of array elements in user's input.
L = size(channel.h_hat, 1); % Length of channel estimator.

%% Convert to baseband and resample the baseband to fs_delay
[p, q] = rat(fs_delay/fs);
baseband = input .* exp(-2j*pi*fc*(0:size(input, 1) - 1).'/fs);
baseband = resample(baseband, p, q);
T = length(baseband);

%% Assign random start point in time (for reproducibility only)
buffer = 20; % extra buffer for extrapolation
T_max = size(channel.h_hat, 3) / fs_time * fs_delay;
if nargin ~= 5
    start = randi([1, T_max - T - L - buffer - 1])
else
    start = varargin{end}
end

%% Convolution and insert the drift
output = zeros(T+buffer+L, M);
baseband = [zeros(L-1, 1); baseband; zeros(L-1, 1)];
channel_time = (0:size(channel.h_hat, 3) - 1) ./ fs_time;
signal_time = ((0:T + L + buffer - 1) + start) ./ fs_delay;
for m = 1:M
    h_hat_m = flip(squeeze(channel.h_hat(:, array_index(m), :)).', 2);
    ir = interp1(channel_time, h_hat_m, signal_time, 'spline');
    if isfield(channel, 'phi_hat')
        for t = 1:T + L - 1
            output(t, m) = ir(t, :) * baseband(t:t+L-1) .* exp(1j*channel.phi_hat(array_index(m), t+start-1));
        end
        % Insert the drift
        drift = channel.phi_hat(array_index(m), (0:T + L + buffer - 1)+start) ./ (2 * pi * fc);
        output(:, m) = interp1(signal_time, output(:, m), signal_time+drift, 'spline');
    elseif isfield(channel, 'theta_hat')
        for t = 1:T + L - 1
            output(t, m) = ir(t, :) * baseband(t:t+L-1) .* exp(1j*channel.theta_hat(array_index(m), t+start-1));
        end
    else
        for t = 1:T + L - 1
            output(t, m) = ir(t, :) * baseband(t:t+L-1);
        end
    end
end

%% Resample to match the original sampling rate and upshift to fc
output_resampled = resample(output, q, p, 'Dimension', 1);
output_resampled = real(output_resampled.*exp(2j*pi*fc*(0:size(output_resampled, 1) - 1).'/fs));

if isfield(channel, 'f_resamp')
    [p2, q2] = rat(channel.f_resamp);
    output_resampled = resample(output_resampled, p2, q2, 'Dimension', 1);
end
output = sqrt(M) ./ sqrt(sum(pwr(output_resampled))) .* output_resampled;

end


function p = pwr(x)
p = mean(abs(x).^2, 1);
end


function validate_inputs(input, fs, index, channel)
% Channel version checking
assert(channel.version >= 1.0, ...
    'The minimum version of the channel matrix is 1.0, and you have %.1f.', ...
    channel.version);

% Get constants
[T, M] = size(input);
T = T / fs;
[~, N, T_max] = size(channel.h_hat);
T_max = T_max / channel.params.fs_time;

% Check the signal lengths
assert(T < T_max, ...
    'Duration of the input signal, %.2fms, should be no larger than %.2fms.', ...
    T*1e3, T_max*1e3);

% Check indices for two modes
assert(length(unique(index)) == length(index));
assert(max(index) <= N, ...
    'array_index must be positive integers and cannot exceed %d.', max(index));
assert(M <= N, 'The maximum supported number of channels is %d.', N);

end

% [EOF]
